h2o-core/src/main/java/water/util/FrameUtils.java

```diff
@@ -111,6 +111,32 @@ public static Frame categoricalEncoder(Frame dataset, String[] skipCols, Model.P
     }
   }
 
+  public static Frame applyTargetEncoder(Model teModel, Frame inputFrame) {
+    Frame transformed = teModel.score(inputFrame);
+    Scope.untrack(transformed.keys());
+
+    // following manipulations with columns is a responsibility of a future pipeline concept. Support for input output columns for the models would be helpful as well.
+    String[] predictorColumns = ArrayUtils.difference(teModel._parms.train().names(), teModel._parms._ignored_columns);
+    final String[] encodedColumns = ArrayUtils.remove(ArrayUtils.remove(predictorColumns, teModel._parms._fold_column), teModel._parms._response_column);
+
+    // We need to rearrange columns ( as some logic relies on response being a last column etc)
+    final String TE_ENCODED_COLUMN_POSTFIX = "_te";
+    Arrays.stream(encodedColumns).forEach(ec -> {
+      transformed
+        .swap(transformed.find(ec), transformed.find(ec + TE_ENCODED_COLUMN_POSTFIX));
+      DKV.put(transformed);
```

Comment:
@sebhrusen 
I see your concerns. Let's discuss and see what could be improved here.
**The main reason why this logic is in utility class is that we need to call it from two different places : ModelBuilder ( when training) and Model ( when scoring).**

Step by step:
1) score creates the frame copy -  I assume we are not going to change it.
2) applyTEIfModelAvailable calls applyTargetEncoder with the copy - yes, but just shallow copy.
3) 
> (what started to track them?) 

I agree, this Scope.track thing is hard to track :)
My understanding is that in non-CV case we are covered by L261 ModelBuilder (see comment L256). In case of CV we enter at L498 and exit at L547 of ModelBuilder. 

But the main "source of inspiration" is a call of `encodeFrameCategoricals` and especially comment at L1371 ModelBuilder. It is using Scope.track and assumes it should be available. The reason why it uses `toDelete` is different - we don't want to delete data too early( but we can, theoretically ). Since `FrameUtils.applyTargetEncoder` calls are analogous to `encodeFrameCategoricals` I assumed we can use Scope as well. 

Also L1173 of Model.java says "Created Vecs have to be deleted by the caller (e.g. Scope.enter/exit)" but I still not sure about having `Scope.untrack(transformed.keys());` in case we call it from Model.java as there is no Scope.enter in place. I will try to think more about this case.

> puts the transformed frame in DKV (or was it just a DKV update after the swaps?)

yes
4) applyTEIfModelAvailable tracks the transformed keys for deletion - i think this is fine. We could have used Scope.track but we can't guaranty that in all use cases Scope.enter is being called.

Reply:
> can't you update the DKV only once after the loop?

@sebhrusen yeap, thanks, we can do that